File: Python/MyEnv.py
import numpy as np
import matplotlib.pyplot as plt
import gymnasium as gym
from gymnasium import spaces

from toolbox import env_mh, INF, SEED


class MyEnv(gym.Env):
    def __init__(self, log_p, max_steps=100):
        super(MyEnv, self).__init__()

        # Target Distribution
        self.log_p = log_p

        # Parameter
        self.max_steps = max_steps
        self.ts = 0  # iteration time

        # Observation specification
        self.observation_space = spaces.Box(low=-INF, high=INF, shape=(1,))

        # Action specification
        self.action_space = spaces.Box(low=-INF, high=INF, shape=(1,))
        self.log_p = lambda x: -x**2/2

        # Store
        self.store_state = [np.array([0.0])]
        self.accetped_status = [True]
        self.store_action = [np.eye(1)]
        self.store_reward = [0.0]

    def step(self, action, policy_cov_func):
        cov_curr = np.diag(np.array([action**2]))

        # MCMC Environment
        state_curr, accepted_status = env_mh(theta_curr=self.state, cov_curr=cov_curr, policy_cov_func=policy_cov_func, log_p=self.log_p)
        self.store_state.append(state_curr)
        self.accetped_status.append(accepted_status)

        # Calculate Reward
        reward = np.power(np.linalg.norm(self.state - state_curr, 2), 2)
        self.store_reward.append(reward)

        # Update Iteration Time
        self.store = state_curr
        self.ts += 1

        # Check for Completion
        terminated = self.ts >= self.max_steps
        truncated = terminated
        if terminated:
            self.reset()

        # Information
        info = {
            "state": state_curr,
            "accepted_status": accepted_status,
            "reward": reward
        }

        return state_curr, reward, terminated, truncated, info

    def reset(self, seed=SEED):
        super().reset(seed=seed)
        self.ts = 0
        self.state = np.array([0.0])  # initialize s_{t}
        self.store_state.append(self.state)
        self.accetped_status.append(True)
        self.store_action.append(np.eye(1))
        self.store_reward.append(0.0)

        # Information
        info = {
            "state": self.state,
            "accepted_status": True,
            "reward": 0.0
        }
        return self.state, info

    def render(self, mode="human"):
        plt.plot(self.store_state)
        plt.show()

File: Matlab/MyEnv.m
classdef MyEnv < rl.env.MATLABEnvironment
    properties
        % Parameter
        sigma = 1;
        nsamples = 1;
        MaxSteps = 100;
        Reward = 0;
        Ts = 0; % iteration time
        State = 0; % state at this time, s_{t}
        OldState = 0; % state at previous state, s_{t-1}

        % Store
        StoreState = {};
        StoreAction = {};
    end

    methods
        function this = MyEnv()
            % Observation specification
            ObservationInfo = rlNumericSpec([1 1]);
            ObservationInfo.Name = 'Obs';
            ObservationInfo.Description = 'Description of the observation';

            % Action specification
            ActionInfo = rlNumericSpec([1 1],'LowerLimit',0,'UpperLimit',inf);
            ActionInfo.Name = 'Act';

            % The following line implements built-in functions of rl.env.VariantEnv
            this = this@rl.env.MATLABEnvironment(ObservationInfo,ActionInfo);

            % Initial State
            this.reset();
        end

        function [NextObs,Reward,IsDone,LoggedSignals] = step(this,Action)
            LoggedSignals = [];

            % Update Action
            this.sigma = Action;

            pdf = @(x) normpdf(x, 4, 1);
            proprnd = @(x) x + this.sigma * randn(1, 1);
            xt = mhsample(this.State,this.nsamples,'pdf',pdf,'proprnd',proprnd,'symmetric',1);

            % Update Obs
            NextObs = xt;
            this.OldState = this.State; % Save s_{t-1}
            this.State = xt; % Update xt in this state

            % Store
            this.StoreState{end+1} = xt;
            this.StoreAction{end+1} = this.sigma;

            % Calculate Reward
            Reward = norm(this.State - this.OldState, 2)^2;

            % Update Iteration Time
            this.Ts = this.Ts + 1;

            % Check for Completion
            IsDone = this.Ts >= this.MaxSteps;

            if IsDone
                this.reset();
            end
        end

        function obs = reset(this)
            this.Ts = 0;
            this.State = 10*rand(1); % initialize s_{t}
            this.OldState = this.State; % initialize s_{t-1}
            obs = this.State;
        end
    end
end
